src/szk.rs

```diff
@@ -0,0 +1,459 @@
+use std::marker::PhantomData;
+///Domain separation tags
+const DST_PROVE_RANDOMNESS: u16 = 0;
+const DST_PROOF_SHARE: u16 = 1;
+const DST_QUERY_RANDOMNESS: u16 = 2;
+const DST_JOINT_RAND_SEED: u16 = 3;
+const DST_JOINT_RAND_PART: u16 = 4;
+const DST_JOINT_RANDOMNESS: u16 = 5;
+
+/// Errors propagated by methods in this module.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum SzkError {
+    /// Calling [`Type::prove`] returned an error.
+    #[error("prove error: {0}")]
+    Prove(String),
+
+    /// Calling [`Type::query`] returned an error.
+    #[error("query error: {0}")]
+    Query(String),
+
+    /// Calling [`Type::decide`] returned an error.
+    #[error("decide error: {0}")]
+    Decide(String),
+
+    /// Generic invalid parameter. This may be returned when a type cannot be constructed.
+    #[error("invalid paramter: {0}")]
+    InvalidParameter(String),
+
+    /// Returned if an FLP operation encountered an error.
+    #[error("Flp error: {0}")]
+    Field(#[from] FlpError),
+
+    /// PRNG error.
+    #[error("prng error: {0}")]
+    Prng(#[from] PrngError),
+
+    /// PRNG error.
+    #[error("codec error: {0}")]
+    Codec(#[from] CodecError),
+
+}
+
+/// Contains an FLP proof share, and if joint randomness is needed, the blind
+/// used to derive it and the other party's joint randomness part.
+pub enum SzkProofShare<'a, const SEED_SIZE: usize, F: FftFriendlyFieldElement>{
+///Leader's proof share is uncompressed.
+///The first Seed is a blind, second is a joint randomness part.
+Leader(Vec<F>, Option<&'a Seed<SEED_SIZE>>, Option<Seed<SEED_SIZE>>),
+///Helper's seed is compressed using the same blind as its joint randomness
+Helper(Option<&'a Seed<SEED_SIZE>>, Option<Seed<SEED_SIZE>>)
+}
+
+/// Verifier type for Szk Proof
+pub type SzkVerifier<F> = Vec<F>;
+
+
+///Main struct encapsulating the shared zero-knowledge functionality.
+/// The type T is the undelrying FLP proof system
+/// P is the XOF used to derive all random coins (it should be indifferentiable
+/// from a random oracle for security.)
+pub struct Szk<T, P, const SEED_SIZE: usize>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    typ: T,
+    dst: [u8; 8],
+    phantom: PhantomData<P>,
+}
+
+impl<T: Type> Szk<T, XofTurboShake128, 16> {
+    /// Create an instance of [`Poplar1`] using [`XofTurboShake128`]. The caller provides the bit length of
+    /// each measurement (`BITS` as defined in [[draft-irtf-cfrg-vdaf-08]]).
+    ///
+    /// [draft-irtf-cfrg-vdaf-08]: https://datatracker.ietf.org/doc/draft-irtf-cfrg-vdaf/08/
+    pub fn new_turboshake128(typ: T, dst: [u8; 8]) -> Self {
+        Szk::new(typ, dst)
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Szk<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    /// Construct an instance of this sharedZK proof system with the underlying FLP
+    /// and a space to hold XOF state for coin derivation.
+    pub fn new(
+        typ: T,
+        dst: [u8; 8]
+    ) -> Self {
```

Comment:
I'd prefer not exposing the bare DST in public API. Also, I think referring to this as "XOF state" is misleading, since that sounds like it's about the internal TurobSHAKE sponge buffer. If the constructor just took in the algorithm ID and stored it, and new DSTs were constructed afresh rather than copying from a template, that should be plenty fast for our purposes, and it would be more readable and misuse-resistant.

Reply:
That's fair, I'll add the algorithm ID to the constructor. The XOF space referenced in the documentation string isn't DST, though, it's the phantom data.

Reply:
Ah, okay. I think we can skip mentioning the `PhantomData` itself in the doc comment, it's mainly an implementation detail to make sure that you can't use multiple XOFs with the same object, and/or to let type inference figure out the XOF type by following dataflow out from the constructor.